eth2/simulation/src/simulation.rs

```diff
@@ -15,69 +18,441 @@ where
 }
 
 impl<T: EthSpec> Simulation<T> {
+    pub fn new() -> Self {
+        Self {
+            store: Store::new(),
+        }
+    }
+
     /// Add a new execution environment, return EE index
-    fn create_execution_environment(a: args::CreateExecutionEnvironment) -> EEIndex {
-        unimplemented!();
+    pub fn create_execution_environment(
+        &mut self,
+        a: simulation_args::CreateExecutionEnvironment,
+    ) -> Result<u64> {
+        // Create internal EE struct from args
+        let ee = ExecutionEnvironment::try_from(a.ee).context(ArgsError)?;
+        let cloned_initial_state = ee.initial_state.clone();
+
+        // Add EE code to beacon chain
+        self.store
+            .current_beacon_state
+            .execution_environments
+            .push(ee)
+            .map_err(|_| Error::MaxLengthExceeded {
+                what: format!("number of execution environments"),
+            })?;
+
+        // For each shard, add the initial state to the shard
+        for shard_state in self.store.current_beacon_state.shard_states.iter_mut() {
+            // Set the initial state of the EE on each ShardState
+            shard_state
+                .execution_environment_states
+                .push(cloned_initial_state)
+                .map_err(|_| Error::MaxLengthExceeded {
+                    what: format!("number of execution environment states"),
+                })?;
+
+            // Each shard should have the same # of ee states as there are EEs
+            assert_eq!(
+                shard_state.execution_environment_states.len(),
+                self.store.current_beacon_state.execution_environments.len()
+            );
+        }
+
+        let ee_index = self.store.current_beacon_state.execution_environments.len() - 1;
+        Ok(ee_index as u64)
     }
 
     /// Add a new shard block containing a list of transactions that need to be executed
     /// Execute all transactions on the appropriate shards / EEs, return ShardBlock index
-    fn create_shard_block(args: args::CreateShardBlock) -> ShardSlot {
-        unimplemented!();
+    pub fn create_shard_block(&mut self, a: simulation_args::CreateShardBlock) -> Result<u64> {
+        // Get the specified ShardState (if it exists)
+        let shard_index = a.shard_index as usize;
+        let shard = Shard::new(a.shard_index);
+        let shard_state = self
+            .store
+            .current_beacon_state
+            .shard_states
+            .get_mut(shard_index)
+            .ok_or(Error::OutOfBounds {
```

Comment:
You should be able to use `context` for a lot of these.

Reply:
I agree, though I personally think the current pattern is more readable than `Snafu`'s `OptionExt` context pattern, simply because it doesn't make use of derived (ie. not easily predictable by most IDEs) context selectors.

But I agree that the `ok_or` could be replaced by:
```
.context(OutOfBounds {
                what: WhatBound::Shard,
                index: shard_index,
            })?;
```
as long as the correct imports were made (`OptionExt` and `OutOfBounds`).

I don't feel too strongly here.  My inclination is to leave this as-is because IMO it's more readable/traceable this way, but also happy to go back and change this if you feel strongly here.  Let me know if so :)